# Анализ данных

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv('dataset_Facebook.csv', sep = ';')
data.head() 

Page total likes    Type  Category  Post Month  Post Weekday  Post Hour  \
0            139441   Photo         2          12             4          3   
1            139441  Status         2          12             3         10   
2            139441   Photo         3          12             3          3   
3            139441   Photo         2          12             2         10   
4            139441   Photo         2          12             2          3   

   Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
0   0.0                       2752                             5091   
1   0.0                      10460                            19057   
2   0.0                       2413                             4373   
3   1.0                      50128                            87991   
4   0.0                       7244                            13594   

   Lifetime Engaged Users  Lifetime Post Consumers  \
0                     178                      109   
1                    1457                     1361   
2                     177                      113   
3                    2211                      790   
4                     671                      410   

   Lifetime Post Consumptions  \
0                         159   
1                        1674   
2                         154   
3                        1119   
4                         580   

   Lifetime Post Impressions by people who have liked your Page  \
0                                               3078              
1                                              11710              
2                                               2812              
3                                              61027              
4                                               6228              

   Lifetime Post reach by people who like your Page  \
0                                              1640   
1                                              6112   
2                                              1503   
3                                             32048   
4                                              3200   

   Lifetime People who have liked your Page and engaged with your post  \
0                                                119                     
1                                               1108                     
2                                                132                     
3                                               1386                     
4                                                396                     

   comment    like  share  Total Interactions  
0        4    79.0   17.0                 100  
1        5   130.0   29.0                 164  
2        0    66.0   14.0                  80  
3       58  1572.0  147.0                1777  
4       19   325.0   49.0                 393

Вычисление среднего, максимального/минимального значений, медианы для всех числовых признаков для всего массива в целом.

In [3]:
l = list(data)
del l[1]
df = pd.DataFrame(columns=['Признак','Среднее значение', 'Максимальное значение', 'Минимальное значение', 'Медиана'])
for i in range(len(l)):
    df = df.append({'Признак': l[i], 'Среднее значение': data[l[i]].mean(),'Максимальное значение': max(data[l[i]]), 'Минимальное значение': min(data[l[i]]), 'Медиана': data[l[i]].median()}, ignore_index=True)
df

Признак  Среднее значение  \
0                                    Page total likes     123194.176000   
1                                            Category          1.880000   
2                                          Post Month          7.038000   
3                                        Post Weekday          4.150000   
4                                           Post Hour          7.840000   
5                                                Paid          0.278557   
6                           Lifetime Post Total Reach      13903.360000   
7                     Lifetime Post Total Impressions      29585.948000   
8                              Lifetime Engaged Users        920.344000   
9                             Lifetime Post Consumers        798.772000   
10                         Lifetime Post Consumptions       1415.130000   
11  Lifetime Post Impressions by people who have l...      16766.376000   
12   Lifetime Post reach by people who like your Page       6585.488000   
13  Lifetime People who have liked your Page and e...        609.986000   
14                                            comment          7.482000   
15                                               like        177.945892   
16                                              share         27.266129   
17                                 Total Interactions        212.120000   

    Максимальное значение  Минимальное значение   Медиана  
0                139441.0               81370.0  129600.0  
1                     3.0                   1.0       2.0  
2                    12.0                   1.0       7.0  
3                     7.0                   1.0       4.0  
4                    23.0                   1.0       9.0  
5                     1.0                   0.0       0.0  
6                180480.0                 238.0    5281.0  
7               1110282.0                 570.0    9051.0  
8                 11452.0                   9.0     625.5  
9                 11328.0                   9.0     551.5  
10                19779.0                   9.0     851.0  
11              1107833.0                 567.0    6255.5  
12                51456.0                 236.0    3417.0  
13                 4376.0                   9.0     412.0  
14                  372.0                   0.0       3.0  
15                 5172.0                   0.0     101.0  
16                  790.0                   0.0      19.0  
17                 6334.0                   0.0     123.5

Мода для числовых признаков.

In [4]:
for i in l:
    print("Мода для признака", i)
    print(data[i].mode())
    print("")

Мода для признака Page total likes
0    136393
dtype: int64

Мода для признака Category
0    1
dtype: int64

Мода для признака Post Month
0    10
dtype: int64

Мода для признака Post Weekday
0    7
dtype: int64

Мода для признака Post Hour
0    3
dtype: int64

Мода для признака Paid
0    0.0
dtype: float64

Мода для признака Lifetime Post Total Reach
0       677
1      2232
2      2645
3      2938
4      3322
5      3358
6      3414
7      3528
8      3754
9      5280
10     5290
11     6692
12     9528
13    13544
14    32208
dtype: int64

Мода для признака Lifetime Post Total Impressions
0     4372
1     6503
2     7004
3     8533
4     8745
5    12735
dtype: int64

Мода для признака Lifetime Engaged Users
0    537
dtype: int64

Мода для признака Lifetime Post Consumers
0    182
1    298
2    319
3    322
4    335
5    374
6    482
7    503
8    513
9    642
dtype: int64

Мода для признака Lifetime Post Consumptions
0    431
1    513
2    599
3    652
4    719
5    730
6    795
7    

Анализ признаков для каждого типа контента.

In [5]:
print('Количество уникальных значений для категориального признака Type.')
a = data['Type'].value_counts()
pd.DataFrame(a)

Количество уникальных значений для категориального признака Type.


Type
Photo    426
Status    45
Link      22
Video      7

In [6]:
photo = data.loc[data['Type'] == 'Photo'].reset_index(drop=True)
status = data.loc[data['Type'] == 'Status'].reset_index(drop=True)
link = data.loc[data['Type'] == 'Link'].reset_index(drop=True)
video = data.loc[data['Type'] == 'Video'].reset_index(drop=True)

Для Type = Photo

In [7]:
dfphoto = pd.DataFrame(columns=['Признак','Среднее значение', 'Максимальное значение', 'Минимальное значение', 'Медиана'])
for i in range(len(l)):
    dfphoto = dfphoto.append({'Признак': l[i], 'Среднее значение': photo[l[i]].mean(),'Максимальное значение': max(photo[l[i]]), 'Минимальное значение': min(photo[l[i]]), 'Медиана': photo[l[i]].median()}, ignore_index=True)
dfphoto

Признак  Среднее значение  \
0                                    Page total likes     122354.171362   
1                                            Category          1.917840   
2                                          Post Month          6.805164   
3                                        Post Weekday          4.107981   
4                                           Post Hour          7.997653   
5                                                Paid          0.280000   
6                           Lifetime Post Total Reach      13137.814554   
7                     Lifetime Post Total Impressions      28994.497653   
8                              Lifetime Engaged Users        818.946009   
9                             Lifetime Post Consumers        690.431925   
10                         Lifetime Post Consumptions       1299.025822   
11  Lifetime Post Impressions by people who have l...      16422.483568   
12   Lifetime Post reach by people who like your Page       6059.103286   
13  Lifetime People who have liked your Page and e...        507.305164   
14                                            comment          7.492958   
15                                               like        182.611765   
16                                              share         27.158768   
17                                 Total Interactions        216.579812   

    Максимальное значение  Минимальное значение   Медиана  
0                139441.0               81370.0  128032.0  
1                     3.0                   1.0       2.0  
2                    12.0                   1.0       7.0  
3                     7.0                   1.0       4.0  
4                    23.0                   1.0       9.0  
5                     1.0                   0.0       0.0  
6                180480.0                 238.0    4675.0  
7               1110282.0                 570.0    8118.5  
8                 11452.0                   9.0     605.5  
9                 11328.0                   9.0     528.5  
10                19779.0                   9.0     827.0  
11              1107833.0                 567.0    5498.0  
12                51456.0                 236.0    3110.0  
13                 3430.0                   9.0     403.0  
14                  372.0                   0.0       3.0  
15                 5172.0                   0.0     100.0  
16                  790.0                   0.0      19.0  
17                 6334.0                   0.0     122.0

In [8]:
for i in l:
    print("Мода для признака", i)
    print(photo[i].mode())
    print("")

Мода для признака Page total likes
0    124940
dtype: int64

Мода для признака Category
0    1
dtype: int64

Мода для признака Post Month
0    10
dtype: int64

Мода для признака Post Weekday
0    7
dtype: int64

Мода для признака Post Hour
0    3
dtype: int64

Мода для признака Paid
0    0.0
dtype: float64

Мода для признака Lifetime Post Total Reach
0       677
1      2232
2      2938
3      3322
4      3358
5      3414
6      3528
7      3754
8      5280
9      5290
10     9528
11    32208
dtype: int64

Мода для признака Lifetime Post Total Impressions
0     4372
1     6503
2     7004
3     8533
4     8745
5    12735
dtype: int64

Мода для признака Lifetime Engaged Users
0    537
dtype: int64

Мода для признака Lifetime Post Consumers
0    298
1    319
2    335
3    374
4    482
5    503
6    513
7    642
dtype: int64

Мода для признака Lifetime Post Consumptions
0    431
1    513
2    599
3    652
4    719
5    730
6    795
7    889
dtype: int64

Мода для признака Lifetime Post Impr

Для Type = Status

In [9]:
dfstatus = pd.DataFrame(columns=['Признак','Среднее значение', 'Максимальное значение', 'Минимальное значение', 'Медиана'])
for i in range(len(l)):
    dfstatus = dfstatus.append({'Признак': l[i], 'Среднее значение': status[l[i]].mean(),'Максимальное значение': max(status[l[i]]), 'Минимальное значение': min(status[l[i]]), 'Медиана': status[l[i]].median()}, ignore_index=True)
dfstatus

Признак  Среднее значение  \
0                                    Page total likes     132647.044444   
1                                            Category          2.022222   
2                                          Post Month          9.066667   
3                                        Post Weekday          4.577778   
4                                           Post Hour          7.244444   
5                                                Paid          0.222222   
6                           Lifetime Post Total Reach      13078.888889   
7                     Lifetime Post Total Impressions      24244.466667   
8                              Lifetime Engaged Users       2040.222222   
9                             Lifetime Post Consumers       1949.555556   
10                         Lifetime Post Consumptions       2838.866667   
11  Lifetime Post Impressions by people who have l...      18664.266667   
12   Lifetime Post reach by people who like your Page       9908.577778   
13  Lifetime People who have liked your Page and e...       1719.844444   
14                                            comment          8.911111   
15                                               like        176.711111   
16                                              share         31.422222   
17                                 Total Interactions        217.044444   

    Максимальное значение  Минимальное значение   Медиана  
0                139441.0              104070.0  135713.0  
1                     3.0                   1.0       2.0  
2                    12.0                   3.0      10.0  
3                     7.0                   1.0       5.0  
4                    15.0                   2.0       9.0  
5                     1.0                   0.0       0.0  
6                 31136.0                3930.0   11096.0  
7                 59964.0                7509.0   20849.0  
8                  6164.0                 128.0    1701.0  
9                  5934.0                  86.0    1599.0  
10                 9237.0                 112.0    2201.0  
11                37849.0                5009.0   17502.0  
12                21352.0                2410.0    8980.0  
13                 4376.0                 101.0    1604.0  
14                   60.0                   0.0       4.0  
15                  859.0                  13.0     150.0  
16                  123.0                   1.0      28.0  
17                 1009.0                  17.0     186.0

In [10]:
for i in l:
    print("Мода для признака", i)
    print(status[i].mode())
    print("")

Мода для признака Page total likes
0    139441
dtype: int64

Мода для признака Category
0    2
dtype: int64

Мода для признака Post Month
0    12
dtype: int64

Мода для признака Post Weekday
0    5
1    7
dtype: int64

Мода для признака Post Hour
0    10
dtype: int64

Мода для признака Paid
0    0.0
dtype: float64

Мода для признака Lifetime Post Total Reach
Series([], dtype: int64)

Мода для признака Lifetime Post Total Impressions
Series([], dtype: int64)

Мода для признака Lifetime Engaged Users
Series([], dtype: int64)

Мода для признака Lifetime Post Consumers
Series([], dtype: int64)

Мода для признака Lifetime Post Consumptions
0    1692
dtype: int64

Мода для признака Lifetime Post Impressions by people who have liked your Page
Series([], dtype: int64)

Мода для признака Lifetime Post reach by people who like your Page
Series([], dtype: int64)

Мода для признака Lifetime People who have liked your Page and engaged with your post
Series([], dtype: int64)

Мода для признака comme

Для Type = Link

In [11]:
dflink = pd.DataFrame(columns=['Признак','Среднее значение', 'Максимальное значение', 'Минимальное значение', 'Медиана'])
for i in range(len(l)):
    dflink = dflink.append({'Признак': l[i], 'Среднее значение': link[l[i]].mean(),'Максимальное значение': max(link[l[i]]), 'Минимальное значение': min(link[l[i]]), 'Медиана': link[l[i]].median()}, ignore_index=True)
dflink

Признак  Среднее значение  \
0                                    Page total likes     116363.181818   
1                                            Category          1.136364   
2                                          Post Month          6.590909   
3                                        Post Weekday          4.272727   
4                                           Post Hour          5.727273   
5                                                Paid          0.272727   
6                           Lifetime Post Total Reach      18544.590909   
7                     Lifetime Post Total Impressions      28725.454545   
8                              Lifetime Engaged Users        342.818182   
9                             Lifetime Post Consumers        292.681818   
10                         Lifetime Post Consumptions        374.090909   
11  Lifetime Post Impressions by people who have l...      11148.590909   
12   Lifetime Post reach by people who like your Page       6544.363636   
13  Lifetime People who have liked your Page and e...        210.545455   
14                                            comment          2.818182   
15                                               like         73.318182   
16                                              share         12.909091   
17                                 Total Interactions         89.045455   

    Максимальное значение  Минимальное значение   Медиана  
0                138895.0               85979.0  115396.0  
1                     3.0                   1.0       1.0  
2                    12.0                   1.0       4.5  
3                     7.0                   1.0       4.5  
4                    14.0                   2.0       4.0  
5                     1.0                   0.0       0.0  
6                 70912.0                1536.0    7422.0  
7                229733.0                3094.0    9773.0  
8                  1374.0                  24.0     244.0  
9                  1106.0                  23.0     205.0  
10                 1345.0                  26.0     290.0  
11                42338.0                2307.0    9034.5  
12                27232.0                1180.0    5039.0  
13                  788.0                  19.0     161.5  
14                   15.0                   0.0       1.5  
15                  379.0                   5.0      37.0  
16                   44.0                   0.0      10.5  
17                  420.0                   6.0      52.5

In [12]:
for i in l:
    print("Мода для признака", i)
    print(link[i].mode())
    print("")

Мода для признака Page total likes
0    138353
dtype: int64

Мода для признака Category
0    1
dtype: int64

Мода для признака Post Month
0    12
dtype: int64

Мода для признака Post Weekday
0    6
dtype: int64

Мода для признака Post Hour
0    3
dtype: int64

Мода для признака Paid
0    0.0
dtype: float64

Мода для признака Lifetime Post Total Reach
Series([], dtype: int64)

Мода для признака Lifetime Post Total Impressions
Series([], dtype: int64)

Мода для признака Lifetime Engaged Users
0    66
dtype: int64

Мода для признака Lifetime Post Consumers
0    322
dtype: int64

Мода для признака Lifetime Post Consumptions
Series([], dtype: int64)

Мода для признака Lifetime Post Impressions by people who have liked your Page
Series([], dtype: int64)

Мода для признака Lifetime Post reach by people who like your Page
Series([], dtype: int64)

Мода для признака Lifetime People who have liked your Page and engaged with your post
Series([], dtype: int64)

Мода для признака comment
0    0
dty

Для Type = Video

In [13]:
dfvideo = pd.DataFrame(columns=['Признак','Среднее значение', 'Максимальное значение', 'Минимальное значение', 'Медиана'])
for i in range(len(l)):
    dfvideo = dfvideo.append({'Признак': l[i], 'Среднее значение': video[l[i]].mean(),'Максимальное значение': max(video[l[i]]), 'Минимальное значение': min(video[l[i]]), 'Медиана': video[l[i]].median()}, ignore_index=True)
dfvideo

Признак  Среднее значение  \
0                                    Page total likes     135014.857143   
1                                            Category          1.000000   
2                                          Post Month          9.571429   
3                                        Post Weekday          3.571429   
4                                           Post Hour          8.714286   
5                                                Paid          0.571429   
6                           Lifetime Post Total Reach      51205.714286   
7                     Lifetime Post Total Impressions     102622.428571   
8                              Lifetime Engaged Users       1707.000000   
9                             Lifetime Post Consumers       1584.714286   
10                         Lifetime Post Consumptions       2600.142857   
11  Lifetime Post Impressions by people who have l...      43149.857143   
12   Lifetime Post reach by people who like your Page      17386.285714   
13  Lifetime People who have liked your Page and e...        979.428571   
14                                            comment         12.285714   
15                                               like        231.428571   
16                                              share         52.142857   
17                                 Total Interactions        295.857143   

    Максимальное значение  Минимальное значение   Медиана  
0                138895.0              126424.0  137893.0  
1                     1.0                   1.0       1.0  
2                    12.0                   6.0      11.0  
3                     6.0                   2.0       3.0  
4                    13.0                   2.0      11.0  
5                     1.0                   0.0       1.0  
6                139008.0               13544.0   30624.0  
7                277100.0               30235.0   56950.0  
8                  3872.0                 459.0    1779.0  
9                  3822.0                 411.0    1643.0  
10                 7327.0                 539.0    2331.0  
11               107502.0               21436.0   30131.0  
12                38720.0                9568.0   14112.0  
13                 2218.0                 363.0     885.0  
14                   23.0                   2.0      17.0  
15                  449.0                  65.0     204.0  
16                  121.0                  13.0      44.0  
17                  550.0                  81.0     271.0

In [14]:
for i in l:
    print("Мода для признака", i)
    print(video[i].mode())
    print("")

Мода для признака Page total likes
0    137893
dtype: int64

Мода для признака Category
0    1
dtype: int64

Мода для признака Post Month
0    11
dtype: int64

Мода для признака Post Weekday
0    2
1    3
dtype: int64

Мода для признака Post Hour
0    11
dtype: int64

Мода для признака Paid
0    1.0
dtype: float64

Мода для признака Lifetime Post Total Reach
Series([], dtype: int64)

Мода для признака Lifetime Post Total Impressions
Series([], dtype: int64)

Мода для признака Lifetime Engaged Users
Series([], dtype: int64)

Мода для признака Lifetime Post Consumers
Series([], dtype: int64)

Мода для признака Lifetime Post Consumptions
Series([], dtype: int64)

Мода для признака Lifetime Post Impressions by people who have liked your Page
Series([], dtype: int64)

Мода для признака Lifetime Post reach by people who like your Page
Series([], dtype: int64)

Мода для признака Lifetime People who have liked your Page and engaged with your post
Series([], dtype: int64)

Мода для признака com

# Самый популярный объект

datanew - новый датасет, в который включены объекты, у которых такие признаки как лайки, комментарии, подписчики и др. выше среднего. А так же только те пользователи, у которых paid = 1, поскольку именно у популярных пользователей чаще всего заказывают рекламу.

In [15]:
f = ((data['Page total likes'] > data['Page total likes'].mean()) & 
     (data['Lifetime Post Total Reach'] > data['Lifetime Post Total Reach'].mean()) & 
     (data['Lifetime Engaged Users'] > data['Lifetime Engaged Users'].mean()) &
     (data['Lifetime Post Total Impressions'] > data['Lifetime Post Total Impressions'].mean()) &
     (data['Lifetime Post Consumers'] > data['Lifetime Post Consumers'].mean()) &
     (data['Lifetime Post Consumptions'] > data['Lifetime Post Consumptions'].mean()) &
     (data['comment'] > data['comment'].mean()) &
     (data['like'] > data['like'].mean()) &
     (data['share'] > data['share'].mean()) &
     (data['Total Interactions'] > data['Total Interactions'].mean()) &
     (data['Lifetime Post reach by people who like your Page'] > data['Lifetime Post reach by people who like your Page'].mean()) &
     (data['Lifetime Post Impressions by people who have liked your Page'] > data['Lifetime Post Impressions by people who have liked your Page'].mean()) &
     (data['Lifetime People who have liked your Page and engaged with your post'] > data['Lifetime People who have liked your Page and engaged with your post'].mean())) 
datanew = data.where(f).dropna()
datanew = datanew
datanew = datanew.where(datanew['Paid'] == 1.0).dropna()
datanew

Page total likes    Type  Category  Post Month  Post Weekday  Post Hour  \
67           138185.0   Photo       1.0        11.0           7.0        3.0   
71           137893.0   Video       1.0        11.0           5.0        3.0   
142          136013.0  Status       2.0        10.0           3.0        2.0   
176          135195.0   Photo       3.0         9.0           6.0        3.0   
223          131808.0  Status       2.0         8.0           6.0       10.0   
243          130791.0   Video       1.0         7.0           3.0       11.0   
244          130791.0   Photo       2.0         7.0           3.0        5.0   
272          127082.0   Photo       1.0         7.0           3.0        3.0   
302          124940.0   Photo       1.0         6.0           5.0       13.0   

     Paid  Lifetime Post Total Reach  Lifetime Post Total Impressions  \
67    1.0                    53456.0                          93790.0   
71    1.0                   100768.0                         220447.0   
142   1.0                    31136.0                          59964.0   
176   1.0                    22304.0                          37159.0   
223   1.0                    20168.0                          35904.0   
243   1.0                    21872.0                          40413.0   
244   1.0                   180480.0                         319133.0   
272   1.0                    76096.0                          94644.0   
302   1.0                    63840.0                         102620.0   

     Lifetime Engaged Users  Lifetime Post Consumers  \
67                   1576.0                    995.0   
71                   2101.0                   1735.0   
142                  6164.0                   5934.0   
176                  1805.0                    984.0   
223                  3370.0                   3244.0   
243                  3872.0                   3822.0   
244                  8072.0                   4010.0   
272                  2889.0                   2495.0   
302                  1382.0                   1229.0   

     Lifetime Post Consumptions  \
67                       1469.0   
71                       2331.0   
142                      9237.0   
176                      1618.0   
223                      5106.0   
243                      7327.0   
244                      6242.0   
272                      4596.0   
302                      1871.0   

     Lifetime Post Impressions by people who have liked your Page  \
67                                             32646.0              
71                                             59658.0              
142                                            35977.0              
176                                            22864.0              
223                                            26881.0              
243                                            24667.0              
244                                           108752.0              
272                                            31096.0              
302                                            32470.0              

     Lifetime Post reach by people who like your Page  \
67                                            14912.0   
71                                            18880.0   
142                                           18048.0   
176                                           13304.0   
223                                           15056.0   
243                                           12920.0   
244                                           51456.0   
272                                           19744.0   
302                                           16480.0   

     Lifetime People who have liked your Page and engaged with your post  \
67                                               884.0                     
71                                               885.0                     
142                                             4376.0                 

Из получившихся девяти объектов сильно выделяется номер 244, особенно по признакам комментарии, репосты, общее количество людей, лайкнушвих страницу, из чего можно сделать вывод, что именно этот объект является самым популярным.